util/limiter/limiter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package limiter
+
+import (
+	_ "embed"
+	"log"
+	"sync"
+	"time"
+
+	"tailscale.com/util/lru"
+)
+
+// Limiter is a keyed token bucket rate limiter.
+//
+// Each key gets its own separate token bucket to pull from, enabling
+// enforcement on things like "requests per IP address". To avoid
+// unbounded memory growth, Limiter actually only tracks limits
+// precisely for the N most recently seen keys, and assumes that
+// untracked keys are well-behaved. This trades off absolute precision
+// for bounded memory use, while still enforcing well for outlier
+// keys.
+//
+// As such, Limiter should only be used in situations where "rough"
+// enforcement of outliers only is sufficient, such as throttling
+// egregious outlier keys (e.g. something sending 100 queries per
+// second, where everyone else is sending at most 5).
+//
+// Each key's token bucket behaves like a regular token bucket, with
+// the added feature that a bucket's token count can optionally go
+// negative. This implements a form of "cooldown" for keys that exceed
+// the rate limit: once a key starts getting denied, it must stop
+// requesting tokens long enough for the bucket to return to a
+// positive balance. If the key keeps hammering the limiter in excess
+// of the rate limit, the token count will remain negative, and the
+// key will not be allowed to proceed at all. This is in contrast to
+// the classic token bucket, where a key trying to use more than the
+// rate limit will get capped at the limit, but can still occasionally
+// consume a token as one becomes available.
+type Limiter[K comparable] struct {
+	// Size is the number of keys to track. Only the Size most
+	// recently seen keys have their limits enforced precisely, older
+	// keys are assumed to not be querying frequently enough to bother
+	// tracking.
```

Comment:
Somewhere we should define the required fields. I think that Size, Max, and RefillInterval must all currently be positive (unless we want to specify some default values so a zero is usable).

Reply:
Separately, it might be nice for us to provide a convenience constructor that takes a size and a rate limit and computes Max and RefillInterval. The calculation isn't hard, but it'd read nicely to get `limiter.New(25, qps)`. YMMV.

Reply:
Hmm, the issue is that Max doesn't have a universally sensible value for consumer who want "just X QPS". The mathematically correct value would be Max=1, but that's almost always too stuttery in real use, and real use wants to allow some amount of bursting before strict enforcement kicks in... But how much? Only the caller knows.

How's this: I added a `QPSInterval(qps)` helper that computes the value of RefillInterval from the given qps target, so an initialization now looks like:

```
l := limiter.Limiter{
  Size: 1234,
  Max: myWellChosenBurstForThisApplication,
  RefillInterval: limiter.QPSInterval(25.0),
}
```

I also find this pleasant because it doesn't give any particular limiter flavor precedence over another, i.e. overdrafting limiters gets built exactly the same way.

WDYT?

(my side rant is that I think x/time/rate's API of asking for a rate and burst ends up being harder to reason about than just telling the user "look, it's a token bucket. It depletes like this, it refills like this" and letting them reason about the underlying implementation, not just some abstract mathematical construct; hence my mild resistance to presenting a rate+burst constructor rather than helpers for configuring a token bucket specifically)

Reply:
I like it.